java/vector/src/main/java/org/apache/arrow/vector/VarBinaryVector.java

```diff
@@ -225,18 +225,17 @@ public void set(int index, NullableVarBinaryHolder holder) {
    */
   public void setSafe(int index, NullableVarBinaryHolder holder) {
     assert index >= 0;
-    fillEmpties(index);
-    BitVectorHelper.setValidityBit(validityBuffer, index, holder.isSet);
-    final int startOffset = getStartOffset(index);
     if (holder.isSet != 0) {
       final int dataLength = holder.end - holder.start;
       handleSafe(index, dataLength);
+      fillHoles(index);
+      final int startOffset = getStartOffset(index);
       offsetBuffer.setInt((index + 1) * OFFSET_WIDTH, startOffset + dataLength);
       valueBuffer.setBytes(startOffset, holder.buffer, holder.start, dataLength);
     } else {
-      handleSafe(index, 0);
-      offsetBuffer.setInt((index + 1) * OFFSET_WIDTH, startOffset);
+      fillEmpties(index + 1);
     }
+    BitVectorHelper.setValidityBit(validityBuffer, index, holder.isSet);
```

Comment:
Is there any special reason for moving this line?

Reply:
yes,
this line requires the validityBuffer to have sufficient capacity, this is guaranteed by `handleSafe`.
notice that this method has two call paths to `handleSafe`, one for the non-null case and the other for the null case. first path calls `handleSafe` directly, the late relies on `fillEmpties` to do so.
no matter which path was taken, by the time the (modified) sets the validity bit, the vector is guaranteed to have sufficient capacity.

Reply:
Sounds good. Thank you for the clarification.